pkg/rmproxy/rmproxy.go

```diff
@@ -375,18 +375,18 @@ func (rmp *RMProxy) UpdateNode(request *si.NodeRequest) error {
 	if rmp.GetResourceManagerCallback(request.RmID) == nil {
 		return fmt.Errorf("received NodeRequest, but RmID=\"%s\" not registered", request.RmID)
 	}
-	go func() {
-		if len(request.Nodes) > 0 {
-			for _, node := range request.Nodes {
-				if len(node.GetAttributes()) == 0 {
-					node.Attributes = map[string]string{}
-				}
-				partition := node.Attributes[siCommon.NodePartition]
-				node.Attributes[siCommon.NodePartition] = common.GetNormalizedPartitionName(partition, request.RmID)
+
+	if len(request.Nodes) > 0 {
```

Comment:
Could you please leave some comments? Otherwise, it looks weird since other requests are sent in parallel ...

Reply:
We can also remove the extra empty check in the node handler loop (iterating an empty map is a no-op and completely safe, plus we really shouldn't have UpdateNode being called with zero nodes).

Reply:
> We can then say that the SI contract is that events are processed in order.

Do we address this by adding comments to code of yunikorn-interface? Or we can add UTs to make sure the order of requests from shim is preserved

Reply:
I don't think we need to do anything really. My intent was just to point out that after these changes we basically have FIFO event processing which should be much more reliable.

Reply:
>  I think for all 3 handlers (app/task/node) we should be good to simply remove the goroutine and execute the code there inline.

Hi @craigcondit, 
Let me conduct a benchmark for app/task before removal. (I assume the "task" you mentioned here refers to allocation)

Below are all the possible functions that could remove their goroutine.
- func (rmp *RMProxy) UpdateApplication(request *si.ApplicationRequest)
- func (rmp *RMProxy) UpdateAllocation(request *si.AllocationRequest) 
- func (rmp *RMProxy) UpdateNode(request *si.NodeRequest) 

Below are functions need to keep their goroutine as they require waiting for a callback:
- func (rmp *RMProxy) RegisterResourceManager(request *si.RegisterResourceManagerRequest, callback api.ResourceManagerCallback) 
- func (rmp *RMProxy) UpdateConfiguration(request *si.UpdateConfigurationRequest) 

> Could you please leave some comments? Otherwise, it looks weird since other requests are sent in parallel ...

@chia7712 If we could remove the goroutine for app/task/node, then the behaviors would aligne and I think the comments will be not necessary. Let the confirm whether we can remove it first.

Reply:
Yes, those three functions are the ones we should update. I think we can even skip the benchmarks. The work being done in these goroutines is extremely trivial, and I believe it was a "premature optimization" to build them this way to begin with. These routines would have to be executed millions of times per second before the effects of removing these goroutines would be visible as anything other than noise. Since these are largely called in response to the receipt of individual K8s events on pods and nodes, such a frequency will never occur.

Reply:
Agree, the heavy frequency will never occur. Will update this commit and run e2e test.

Reply:
e2e test passed: https://github.com/chenyulin0719/yunikorn-k8shim/actions/runs/7587158325

> We can also remove the extra empty check in the node handler loop (iterating an empty map is a no-op and completely safe, plus we really shouldn't have UpdateNode being called with zero nodes).

Also remove other empty list check in UpdateApplication/UpdateAllocation.